Level 3

mariya-goncharenko opened the pull request:
Дипломная  работа готова к проверке. Обе части отправлены на сервер. Функционал сайта работает. 

Почему-то добавились коммиты из предыдущего реквеста, работа над 4 частью начинается с коммита 05c61d171afba4a65f9a5c5dfb4625c11d0a5c1a, названного как "Начало работы над 4 этапом дипломной работы." от 28.06.2023.

mariya-goncharenko committed:
Начало работы над версткой. Установка шрифтов и начало работы над структурой.

mariya-goncharenko committed:
Верстка главной страницы сайта.

mariya-goncharenko committed:
Верстка регистрации и авторизации.

Нужно поправить меню.

mariya-goncharenko committed:
Страница профиля. Убрала меню в регистрации/авторизации.

mariya-goncharenko committed:
Продолжение работы над вёрсткой.

mariya-goncharenko committed:
Изменен нейминг, проверка БЭМ.

mariya-goncharenko committed:
Добавлена страница с фильмами.

mariya-goncharenko committed:
Добавила сохраненные фильмы

mariya-goncharenko committed:
json

mariya-goncharenko committed:
Исправление недочётов. Первый прогон.

mariya-goncharenko committed:
Прогон через валидаторы.

mariya-goncharenko committed:
Валидация HTML.

mariya-goncharenko committed:
Подгон по пикселю.

mariya-goncharenko committed:
Вёрстка готова к отправке.

mariya-goncharenko committed:
Первая итерация правок после ревью.

mariya-goncharenko committed:
Изменение чекбокса

mariya-goncharenko committed:
Начало работы над 4 этапом дипломной работы.

mariya-goncharenko committed:
Добавлены методы API.

mariya-goncharenko committed:
Добавила функции фильтрации и константы для отображения карточек.

mariya-goncharenko committed:
Добавлен контекст и поп-ап формы.

mariya-goncharenko committed:
Обновлены обращения к АПИ

mariya-goncharenko committed:
Прописана логика в App.js.

mariya-goncharenko committed:
Прописана логика работы приложения.

Перенос логики с чернового варианта на ВМ. Подготовка проекта к залитию на ВМ и тестам (Проверка работоспособности, правильность CSS и прочего).

mariya-goncharenko committed:
Исправление логики профиля.

mariya-goncharenko committed:
Попытки исправить валидацию форм.

mariya-goncharenko committed:
Подготовленный к сдаче проект.

mariya-goncharenko committed:
Исправлены небольшие отклонения вёрстки.

mariya-goncharenko committed:
Готовая дипломная работа.

mariya-goncharenko committed:
Добавлена ссылка на реквест.

irinamozes commented:
Здравствуйте, Мария. 
 
Не было возможности зарегистрироваться на Вашем сайте: под полем имени всегда горит сообщение об ошибке, что бы в него ни было введено (см. снимок), и  кнопка "Зарегистрироваться" не становится активной. Проверьте функционал валидации форм аутенификации и регулярные выражения для полей ввода. Если значение поля должно соответствовать какому-то нестандартному формату, этот формат для пользователя нужно указать. 
![Снимок экрана в 2023-07-09 20-41-35](https://github.com/mariya-goncharenko/movies-explorer-frontend/assets/19558891/6b47e5ea-4d96-497d-8b7a-ca3c0ddb021b)

Протестируйте, пожалуйста, свой проект перед следующей отправкой на ревью.

Текущая итерация засчитываться не будет без потери количества итераций.

irinamozes commented:
@mariya-goncharenko
Извините, пожалуйста. В резолюции по работе по ошибке указано не Ваше имя. Должно быть Мария.

mariya-goncharenko commented:
@irinamozes 

Добрый вечер, Ирина! Подскажите пожалуйста с какого браузера вы просматривали сайт? 
Протестировала сейчас с Яндекс.Браузера и с Хрома, всё работает. 
Возможно я не заметила эту ошибку из-за отсутствия у меня нужного браузера.

mariya-goncharenko commented:
Попробовала ещё с Duck Duck Go, на андроиде.
Проблему обнаружила только в Сафари, попросив друзей зарегистрироваться. 
Теперь у меня возник вопрос, как мне тестировать сайт, если у меня нет возможности просматривать изменения в Сафари на постоянной основе?
Буду рада если вы подскажете как поступить в такой ситуации.

irinamozes commented:
@mariya-goncharenko
 
Добрый вечер! Сайт я просматривала в Яндекс Браузере Версия 23.5.1.575 beta (64-bit). Beta-версия,потому что у меня ОС Linux, но эта версия новая и последняя. Сейчас я открыла Ваш сайт в Firefox. Там то же самое - сообщение об ошибке и кнопка "Зарегистрироваться" не становится активной. (см. снимок). Я бы советовала не использовать регулярное выражение для поля имени. Для этого поля достаточно указания минимальной и максимальной длины. Символы в него можно разрешить вводить любые.
![Снимок экрана в 2023-07-09 22-10-04](https://github.com/mariya-goncharenko/movies-explorer-frontend/assets/19558891/176683d5-a12f-4080-a832-d2d5008763c7)

mariya-goncharenko commented:
Спасибо большое! Исправлю все недочеты и пришлю новую версию на проверку.

mariya-goncharenko committed:
Подготовка проекта к ревью.

Убрала паттерн для пользователя. Поправила всплывашки и нейминг констант. Проверка кода на отступы.

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 98.
 
## У Вас отличный дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Руководство Typescript
http://typescript-lang.ru/docs/

TypeScript в деталях.
https://typescript-definitive-guide.ru/book/chapters

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
